File: sort-colors/Solution75.java
// https://leetcode.com/problems/sort-colors/

import java.util.Arrays;

class Solution75 {

    // T: O(n), S: O(1)
    public void sortColors(int[] nums) {
        int cnt0 = 0, cnt1 = 0, cnt2 = 0;
        for (int n : nums) {
            if (n == 0) {
                cnt0++;
            } else if (n == 1) {
                cnt1++;
            } else if (n == 2) {
                cnt2++;
            }
        }
        int i = 0;
        while (cnt0 > 0) {
            nums[i++] = 0;
            cnt0--;
        }
        while (cnt1 > 0) {
            nums[i++] = 1;
            cnt1--;
        }
        while (cnt2 > 0) {
            nums[i++] = 2;
            cnt2--;
        }
    }

    // T: O(n), S: O(1)
    public void sortColorsOnePass(int[] nums) {
        int p0 = 0, p1 = 0, p2 = nums.length - 1;
        while (p1 <= p2) {
            if (nums[p1] == 0) {
                swap(nums, p0, p1);
                p0++;
                p1++;
            } else if (nums[p1] == 1) {
                p1++;
            } else {
                swap(nums, p2, p1);
                p2--;
            }
        }
    }

    private static void swap(int[] arr, int i, int j) {
        if (i != j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void main(String[] args) {
        var sol = new Solution75();
        var nums1 = new int[] { 2, 0, 2, 1, 1, 0 };
        var nums2 = new int[] { 2, 0, 1 };
        var nums3 = new int[] { 0, 2, 1 };
        sol.sortColorsOnePass(nums1);
        sol.sortColorsOnePass(nums2);
        sol.sortColors(nums3);
        System.out.println(Arrays.toString(nums1)); // [0,0,1,1,2,2]
        System.out.println(Arrays.toString(nums2)); // [0,1,2]
        System.out.println(Arrays.toString(nums3)); // [0,1,2]
    }

}

File: sort-colors/solution75.py
# https://leetcode.com/problems/sort-colors/

from typing import List


class Solution75:

    # T: O(n), S: O(1)
    def sortColors(self, nums: List[int]) -> None:
        cnt0, cnt1, cnt2 = 0, 0, 0
        for n in nums:
            if n == 0:
                cnt0 += 1
            elif n == 1:
                cnt1 += 1
            elif n == 2:
                cnt2 += 1
        i = 0
        while cnt0 > 0:
            nums[i] = 0
            i += 1
            cnt0 -= 1
        while cnt1 > 0:
            nums[i] = 1
            i += 1
            cnt1 -= 1
        while cnt2 > 0:
            nums[i] = 2
            i += 1
            cnt2 -= 1

    # T: O(n), S: O(1)
    def sortColorsOnePass(self, nums: List[int]) -> None:
        p0, p1, p2 = 0, 0, len(nums) - 1
        while p1 <= p2:
            if nums[p1] == 0:
                nums[p0], nums[p1] = nums[p1], nums[p0]
                p0 += 1
                p1 += 1
            elif nums[p1] == 1:
                p1 += 1
            else:
                nums[p2], nums[p1] = nums[p1], nums[p2]
                p2 -= 1


def main():
    sol = Solution75()
    nums1 = [2, 0, 2, 1, 1, 0]
    nums2 = [2, 0, 1]
    nums3 = [0, 2, 1]
    sol.sortColorsOnePass(nums1)
    sol.sortColorsOnePass(nums2)
    sol.sortColors(nums3)
    print(nums1)  # [0,0,1,1,2,2]
    print(nums2)  # [0,1,2]
    print(nums3)  # [0,1,2]


if __name__ == "__main__":
    main()
